Adding python3 test module for TC_ACL_2_5

j-ororke opened the pull request:
- Added python3 test module for ACL 2 5
- Removed Test_TC_ACL_2_5 yaml test script

- This uses base branch of Amine's PR https://github.com/project-chip/connectedhomeip/pull/37755 to create this test module: This is a follow-up PR to that PR from Amine, once Amine's PR is merged this one can come out of WIP state.

### Testing
WSL - Windows Linux Subsystem

Alami-Amine committed:
Testing removing one of the template functions

Alami-Amine committed:
WriteClient: Encoding as many items as possible into a single list as part of ReplaceAll item operation

Alami-Amine committed:
fixing failures

Alami-Amine committed:
Fix to Error Handling in User-Label Cluster: return RESOURCE_EXHAUSTED when the number of items in list exceeds kMaxUserLabelListLength

Alami-Amine committed:
chip-repl: support for preencoded attributes

Alami-Amine committed:
Fixing ACL 2.3 and 2.5

j-ororke committed:
Added TC_ACL_2_5 python3 test:
- Added initial python3 test for TC_ACL_2_5

j-ororke committed:
Removed Test_TC_ACL_2_5.yaml:
- Removed Test_TC_ACL_2_5 yaml test script as we are adding python3 test module

j-ororke committed:
Updated TC_ACL_2_5 python test module:
- Removed unneccessary EventCallback class as not used in test

j-ororke committed:
Updating TC_ACL_2_5 python3 test module:
- Renaming returned value for desc_TC_ACL_2_5()

bzbarsky-apple commented:
Converted this to draft, until https://github.com/project-chip/connectedhomeip/pull/37755 is merged.

j-ororke committed:
style: format TC_ACL_2_5.py with autopep8 and isort

Alami-Amine commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,600 @@
+#
+#    Copyright (c) 2023 Project CHIP Authors
```
should we put 2025?

j-ororke replied:
Yes, good catch, that should definitely be 2025 now. Thank you!

Alami-Amine commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,600 @@
+                     "Result is SUCCESS, value is list of AccessControlExtensionChanged containing 1 element"),
+            TestStep(6, "TH1 writes DUT Endpoint 0 AccessControl cluster Extension attribute, value is list of AccessControlExtensionStruct containing 1 element", "Result is SUCCESS"),
+            TestStep(7, "TH1 reads DUT Endpoint 0 AccessControl cluster AccessControlExtensionChanged events",
+                     "Result is SUCCESS, value is list of AccessControlExtensionChanged containing at least 2 new elements"),
```
As discussed
```suggestion
                     "Result is SUCCESS, value is list of AccessControlExtensionChanged containing 1 element"),
```

Alami-Amine replied:
will it be more informative to add "1 element with ChangeType = Changed" ? If yes, we can do the same for Step 5 above  "1 element with ChangeType = Added"

Alami-Amine commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,600 @@
+        self.step(4)
+        # Create an extension with a test string
+        # Use a properly formatted byte string - not a string that looks like bytes
+        extension_data = bytes.fromhex('1718')
```
Should we name this to show the "type of data", as is done in YAML where this was called `D_OK_EMPTY` ? It will mak tracking it easier and we can have the same name across other tests (such as 2_7)

j-ororke replied:
Thank you, good point, I have update this, it will be much easier to track that way.

Alami-Amine commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,600 @@
+        self.step(6)
+        # Create a new extension with different data to replace the existing one
+        # Use properly formatted binary data
+        new_extension_data = bytes.fromhex(
```
same as for `extension_data` above, `D_OK_SINGLE`  was the name used in YAML tests. It will make tracking it easier and we can have the same name across other tests (such as 2_7)

j-ororke replied:
Also good point, I have now changed this to be easier to track.

Alami-Amine commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,600 @@
+
+                    if hasattr(event_data, 'Data') and hasattr(event_data.Data, 'changeType'):
+                        # Check for remove event
+                        # Review: Does this event have to be of kRemoved, or can it be kChanged?
```
Actually it can only be KChanged; we only have one Possible entry in ACL Extensions, and in Step 6, we replaced that extension.
Before my changes, we used to first "remove" the extension then "Append" it, now, we just "update/change" it directly

j-ororke replied:
Awesome, thank you, then I have updated this to only look for the changed event during test step 7

Alami-Amine commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,600 @@
+                            self.print_step(f"Found Remove/Change event", "")
+
+                        # Check for add event with matching data
+                        elif (event_data.Data.changeType == Clusters.AccessControl.Enums.ChangeTypeEnum.kAdded and
```
as discussed, we shouldn't have a `ChangeTypeEnum.kAdded` in this step,
the key is to find a way to add `--event-min 6` to this command so we only see the `ChangeTypeEnum.kChanged` entry and not the leftover event

j-ororke replied:
Thank you, 
The below eventNumberFilter for the ReadEvent you mentioned sorted this out quick and we now only get the changed event back for test step 7 as expected, as eventNumberFilter=6 is the ChangeTypeEnum.kChanged event.

Alami-Amine commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,600 @@
+        too_long_extension = Clusters.AccessControl.Structs.AccessControlExtensionStruct(
+            data=too_long_data)
+
+        # This should fail with RESOURCE_EXHAUSTED
```
```suggestion
        # This should fail with CONSTRAINT_ERROR
```

Alami-Amine commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,600 @@
+                self.print_step(f"Examining event", str(event_data))
+
+                if hasattr(event_data, 'Data') and hasattr(event_data.Data, 'changeType'):
+                    # Only check add events
```
I think its better to check that we don't have ANY events at all, not just `kAdded` events.
Maybe adding `eventNumberFilter` in `ReadEvent` above will do it? we shouldn't have any events with EventNumber > 6

j-ororke replied:
Thank you, that was a great idea, and it worked really well it appears. setting eventNumberFilter=7 returned an empty list when reading the events. So, then no events since the changed event above for event number 6 has been noticed.

Alami-Amine commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,600 @@
+                f"Failed to write multiple extensions due to other error: {e}")
+
+        self.step(11)
+        # Verify no event was generated for the second extension
```
```suggestion
        # Verify no event was generated at all, since the whole extensions list was rejected.
```

Alami-Amine commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,600 @@
+
+            found_valid_event = False
+
+            # Ensure no events have the second extension data
```
```suggestion
            # Ensure that no events were generated at all
```

Alami-Amine commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,600 @@
+            events_response = await self.default_controller.ReadEvent(
+                self.dut_node_id,
+                events=[(0, acec_event)],
+                fabricFiltered=True
```
idem as in step 9, add `eventNumberFilter` making sure that we check all events with event number > 6

Alami-Amine replied:
I think if you do that, you can then just check that `events_response` is 0 and you wouldn't need the `for` loop in line 479

j-ororke replied:
Thank you, that worked way better and cleaner than the for loop that had been there before!

j-ororke committed:
Apply suggestions from code review

Committing suggestions from Amine, which are great changes/additions and helped improve the code here, thank you Amine!

Co-authored-by: Amine Alami <43780877+Alami-Amine@users.noreply.github.com>

j-ororke committed:
Updating TC_ACL_2_5.py test module:
- Started using eventNumberFilter var for ReadEvent.
- Removed looking for add event in test step 7
- Changed variable names for test vars to match with yaml scripts.

j-ororke committed:
Updating TC_ACL_2_5 python3 test module:
- Resolving linting errors

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updated TC_ACL_2_5 python3 test module:
- Resolving linting errors

j-ororke committed:
Updating TC_ACL_2_5 python3 test module:
- Resolving linting errors

restyled-commits committed:
Restyled by isort

Alami-Amine commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,510 @@
+
+            # Extract events from the response
+            events = events_response
+            self.print_step(f"Found {len(events)} initial events", "")
```
Could you please add here an assertion that there are no events:
`asserts.assert_equal(len(events), 0, "There should be no events found")`

j-ororke replied:
Updated to include the assertion after the print statement.

Alami-Amine commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,510 @@
+            events = events_response
+            self.print_step(f"Found {len(events)} initial events", "")
+
+            for event_data in events:
```
and I guess we can remove this, since we should have no events

j-ororke committed:
Updating TC_ACL_2_5 python3 test module:
- Conforming to new test steps in PR from Amine PR #[5061](https://github.com/CHIP-Specifications/chip-test-plans/pull/5061)

j-ororke committed:
Updating TC_ACL_2_5 python3 test module:
- Resolving issue with copyright date, should be 2025, but was 2023
- Added assertion check for length of events being 0 during test step 3
- Removed for loop in test step 3 as not needed with new assertion

j-ororke commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,507 @@
+            
+            # Extract events from the response
+            events = events_response
+            self.print_step(f"Found {len(events)} initial events", "")
```
No problem, updated to replace self.print_step() with logging.info() method.

j-ororke committed:
Updating TC_ACL_2_5 python3 test module:
- Updated to remove direct read of events to using EventChangeCallback with wait_for_event_report() for subscription method
- Removed exceptions as no longer needed
- Removed print_step() with logging.info()

j-ororke commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,507 @@
+        await asyncio.sleep(5)  # Increased wait time
+
+        self.step(5)
+        # Read the events directly instead of relying on subscription
```
Good question, I have changed to using the subscription method EventChangeCallback with wait_for_event_report(), this method makes the code much cleaner for this test.

Alami-Amine commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,316 @@
+            self.dut_node_id,
+            events=[(0, acec_event)],
+            fabricFiltered=True,
+            eventNumberFilter=7
```
Would it be a good idea to read eventNumber of last received event, and then just use it here in the filter like "LastReceivedEventNumber + 1", it will be more maintainable that just the number? sorry for giving wrong tip before, I hadn't thought of this way, until I saw it in the YAML.

j-ororke replied:
Thank you, this has been updated

j-ororke committed:
chore: sync WriteClient.h, WriteClient.cpp, user-label-server.cpp, TestWriteInteraction.cpp, and TestWriteChunking.cpp with upstream master before rebase

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Resolving linting errors

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updated TC_ACL_2_5 python3 test module:
- Cleanup

restyled-commits committed:
Restyled by autopep8

cecille commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,298 @@
+        await events_callback.start(self.default_controller, self.dut_node_id, 0)
+
+        self.step(4)
+        # Create an extension with a test string
```
```suggestion
        # Create an extension with a test string (test string specified by test plan)
        # 1718 = octstr of length 2 containing valid TLC for a top level anonymous list (empty)
```

cecille commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,298 @@
+                             "LatestValue.FabricIndex should be the current fabric index")
+        asserts.assert_equal(event_data.fabricIndex,
+                             f1,
+                             "FabricIndex should be the current fabric index")
```
test plan specifies that the event should be read. To be fair, this way makes MUCH more sense and I would argue is better, but can you also add a read in here and ensure that the read value matches the value received in the event subscription? Same below. Follow up is fine. Follow up might be preferable, actually.

j-ororke replied:
No problem, I have created a short follow-up task for this here: https://github.com/project-chip/matter-test-scripts/issues/547

cecille commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,298 @@
+                             "FabricIndex should be the current fabric index")
+
+        self.step(6)
+        # Create a new extension with different data to replace the existing one
```
```suggestion
        # Create a new extension with different data to replace the existing one
        # Value is from test plan and is a valid top level anonymous list containing one element.
```

cecille commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,298 @@
+                             "FabricIndex should be the current fabric index")
+
+        self.step(8)
+        # Try to write an extension that exceeds max length (128 bytes)
```
```suggestion
        # Try to write an extension that exceeds max length (128 bytes)
        # Value is from test plan and is a valid top-level anonymous list with a length of 129 (too long)
```

cecille commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,298 @@
+            self.dut_node_id,
+            events=[(0, acec_event)],
+            fabricFiltered=True,
+            eventNumberFilter=7
```
This event number shouldn't be hardcoded - there's no guarantee this is the 7th event, especially if the device is not factory reset directly before this test. This should come from the last event.

j-ororke replied:
Sorry, this has been updated here now, Amine had also mentioned this for test step 11 below. 
Updated now to using a dynamic value set from the max EventNumber noticed in AccessControlExtensionStruct event path.

cecille commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,298 @@
+        logging.info(f"Events response {str(events_response)}")
+
+        # Extract events from the response
+        events = events_response
```
same q as above

cecille replied:
and below

j-ororke replied:
Ah yes, sorry as mentioned above this was initially just looking at the events_response[0] and assigned to events var, but found it is not needed or logical, and those var set lines are now removed.

j-ororke committed:
Applying suggestions from code review from Cecille

Co-authored-by: C Freeman <cecille@google.com>

j-ororke commented on src/python_testing/TC_ACL_2_5.py:
```diff
@@ -0,0 +1,298 @@
+
+        # Set up event subscription before making changes
+        logging.info("Setting up event subscription...")
+        events_callback = EventChangeCallback(Clusters.AccessControl)
```
Added to test steps, will create a test plan PR to add this in as well

j-ororke committed:
Updating TC_ACL_2_5 test module:
- Removed unneeded events var
- Set value for eventNumberFilter to be dynamic
- Created new follow-up task for test step 5 for validating read events matches subscription events data

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Merge branch 'master' into update_TC_ACL_2_5_test

Alami-Amine commented:
- Backward Compatibility: Passed Tests when run against `chip-all-clusters-app` from Matter 1.0
